### DISCUSSION TITLE
Feature Request - Branch Protection Rules for Source Branches

### DISCUSSION THREAD

**Discussion by brosmar** [community/community#102765] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

Dear GitHub Team,

I am writing to suggest a new feature that I believe would significantly enhance the utility and control of branch management in GitHub.

Currently, GitHub's branch protection rules allow us to control the merging into a protected branch. However, there is no built-in functionality to control or restrict merging from a source branch based on its status.

In certain scenarios, it would be beneficial to prevent merging from branches that do not meet specific criteria. For example, a branch that has failing tests, is behind the main branch, or does not adhere to certain predefined rules should not be merged into another branch.

I propose the addition of branch protection rules for source branches. These rules could include conditions like passing all checks, being up-to-date with the target branch, or any other custom conditions defined by the repository administrators.

This feature would provide more granular control over the code quality and integrity, reduce human error, and streamline the code review process. It would also complement the existing branch protection rules and further enhance GitHub's robust version control capabilities.

I understand that implementing new features depends on various factors, but I believe this feature would be a valuable addition to many GitHub users.

Thank you for considering my suggestion. I look forward to seeing continuous improvements to the GitHub platform.

Best regards,

Martin

**Comment by AndreaGriffiths11** [comment]
Thanks for the feedback @brosmar, I wonder if this is something that ruleset can help with? 🤔 Rulesets are used to enforce certain workflows or requirements before branches can be merged, like requiring pull request reviews or passing status checks. https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets#about-rulesets

**Comment by brosmar** [comment]
Maybe I did Somehow, I have a knot in my head. I don't know what prompted me to make this request. I need to think about it again. Therefore, I will close the post.
